pkg/sbom/cyclonedx/cyclonedx.go

```diff
@@ -0,0 +1,412 @@
+)
+
+type TrivyBOM struct {
+	cdx.BOM
+}
+
+func (b TrivyBOM) parseComponents() (string, *ftypes.OS, map[string]*ftypes.Application, map[string]cdx.Component, error) {
+
+	var osBOMRef string
+	appMap := make(map[string]*ftypes.Application)
+	libMap := make(map[string]cdx.Component)
+
+	var os *ftypes.OS
+	for _, component := range *b.Components {
+		switch component.Type {
+		case cdx.ComponentTypeOS:
+			osBOMRef = component.BOMRef
+			os = b.OS(component)
+		case cdx.ComponentTypeApplication:
+			appMap[component.BOMRef] = b.Application(component)
+		case cdx.ComponentTypeLibrary:
+			if t := getProperty(component.Properties, PropertyType); t != "" {
+				// If type property exists, it is Application.
+				app := ftypes.Application{
+					Type:     t,
+					FilePath: getProperty(component.Properties, PropertyFilePath),
+				}
+
+				pkg, err := b.Package(component)
+				if err != nil {
+					return "", nil, nil, nil, xerrors.Errorf("failed to parse package: %w", err)
+				}
+				app.Libraries = []ftypes.Package{*pkg}
+				appMap[component.BOMRef] = &app
+			} else {
+				// If it isn't application component, it is library.
+				libMap[component.BOMRef] = component
+			}
+		}
+	}
+
+	return osBOMRef, os, appMap, libMap, nil
+}
+
+func (b TrivyBOM) Aggregate(libs []cdx.Component) ([]ftypes.Application, error) {
+	appsMap := map[string]*ftypes.Application{}
+	for _, lib := range libs {
+		p, err := packageurl.FromString(lib.PackageURL)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to parse purl from string: %w", err)
+		}
+
+		app, ok := appsMap[p.Type]
+		if !ok {
+			app = &ftypes.Application{
+				Type:     purl.Type(p.Type),
+				FilePath: p.Type,
+			}
+			appsMap[p.Type] = app
+		}
+		pkg, err := b.Package(lib)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to parse purl to package: %w", err)
+		}
+
+		app.Libraries = append(app.Libraries, *pkg)
+	}
+
+	var apps []ftypes.Application
+	for _, app := range appsMap {
+		apps = append(apps, *app)
+	}
+	return apps, nil
+}
+
+func typeFromComponent(c cdx.Component) (string, error) {
+	p, err := packageurl.FromString(c.PackageURL)
+	if err != nil {
+		return "", xerrors.Errorf("failed to parse purl: %w", err)
+	}
+	return purl.Type(p.Type), nil
+}
+
+func (b TrivyBOM) OS(component cdx.Component) *ftypes.OS {
+	return &ftypes.OS{
+		Family: component.Name,
+		Name:   component.Version,
+	}
+}
+
+func (b TrivyBOM) Application(component cdx.Component) *ftypes.Application {
+	return &ftypes.Application{
+		Type:     getProperty(component.Properties, PropertyType),
+		FilePath: component.Name,
+	}
+}
+
+func (b TrivyBOM) Package(component cdx.Component) (*ftypes.Package, error) {
+	pkg, err := purl.FromString(component.PackageURL)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to parse purl: %w", err)
+	}
+	pkg.Ref = component.BOMRef
+
+	if component.Licenses != nil {
+		for _, license := range *component.Licenses {
+			pkg.License = license.Expression
```

Comment:
Looks like it overwrites a license. To avoid future confusion, we should join all the licenses.